tiatoolbox/tools/stainextract.py

```diff
@@ -14,35 +14,207 @@
         self.stain_matrix = stain_matrix
         if self.stain_matrix.shape != (2, 3) and self.stain_matrix.shape != (3, 3):
             raise Exception("Stain matrix must be either (2,3) or (3,3)")
 
     def get_stain_matrix(self, _):
+        """Get the user defined stain matrix.
+
+        Returns:
+            ndarray: user defined stain matrix.
+
+        """
         return self.stain_matrix
 
 
 class RuifrokExtractor:
     """Get the stain matrix as defined in:
 
-    A.C. Ruifrok & D.A. Johnston 'Quantification of histochemical staining
-    by color deconvolution'. Analytical and quantitative cytology and histology
-    / the International Academy of Cytology and American Society of Cytology.
+    Ruifrok, Arnout C., and Dennis A. Johnston. "Quantification of
+    histochemical staining by color deconvolution." Analytical and
+    quantitative cytology and histology 23.4 (2001): 291-299.
+
+    This class contains code inspired by StainTools
+    [https://github.com/Peter554/StainTools] written by Peter Byfield.
+
+    Examples:
+        >>> from tiatoolbox.tools.staiextract import RuifrokExtractor()
+        >>> extractor = RuifrokExtractor()
+        >>> stain_matrix = extractor.get_stain_matrix(img)
 
     """
 
     @staticmethod
     def get_stain_matrix(_):
+        """Get the pre-defined stain matrix.
+
+        Returns:
+            ndarray: pre-defined  stain matrix.
+
+        """
         return np.array([[0.65, 0.70, 0.29], [0.07, 0.99, 0.11]])
+
+
+class MacenkoExtractor:
+    """Get the stain matrix as defined in:
+
+    Macenko, Marc, et al. "A method for normalizing histology
+    slides for quantitative analysis." 2009 IEEE International
+    Symposium on Biomedical Imaging: From Nano to Macro. IEEE, 2009.
+
+    This class contains code inspired by StainTools
+    [https://github.com/Peter554/StainTools] written by Peter Byfield.
+
+    Examples:
+        >>> from tiatoolbox.tools.staiextract import MacenkoExtractor()
+        >>> extractor = MacenkoExtractor()
+        >>> stain_matrix = extractor.get_stain_matrix(img)
+
+    """
+
+    @staticmethod
+    def get_stain_matrix(img, luminosity_threshold=0.8, angular_percentile=99):
+        """Stain matrix estimation.
+
+        Args:
+            img (ndarray): input image used for stain matrix estimation
+            luminosity_threshold (float): threshold used for tissue area selection
+            angular_percentile (int):
+
+        Returns:
+            ndarray: estimated stain matrix.
+
+        """
+        img = img.astype("uint8")  # ensure input image is uint8
+        # convert to OD and ignore background
+        tissue_mask = get_luminosity_tissue_mask(
+            img, threshold=luminosity_threshold
+        ).reshape((-1,))
+        img_od = convert_RGB2OD(img).reshape((-1, 3))
+        img_od = img_od[tissue_mask]
+
+        # eigenvectors of cov in OD space (orthogonal as cov symmetric)
+        _, V = np.linalg.eigh(np.cov(img_od, rowvar=False))
```

Comment:
Can we use a variable which is self-explanatory instead of just V?

Reply:
Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/45/commits/0f01e044ef7fba19b49dcaf7c06017370a796eff